src/txn/txn.c

```diff
@@ -177,16 +177,19 @@ __wt_txn_get_snapshot(WT_SESSION_IMPL *session)
     /*
      * Include the checkpoint transaction, if one is running: we should ignore any uncommitted
      * changes the checkpoint has written to the metadata. We don't have to keep the checkpoint's
-     * changes pinned so don't including it in the published pinned ID.
+     * changes pinned so don't go including it in the published pinned ID.
+     *
+     * We can assume that if a function calls without intention to publish then it is the special
+     * case of checkpoint calling it twice. In which case do not include the checkpoint id.
      */
-    if ((id = txn_global->checkpoint_txn_shared.id) != WT_TXN_NONE) {
+    if ((id = txn_global->checkpoint_txn_shared.id) != WT_TXN_NONE && publish) {
         txn->snapshot[n++] = id;
```

Comment:
I think it's a bad idea to assume `publish => called from checkpoint`.  One day we'll use this from somewhere else and get burned.  Can you instead do:

```c
if ((id = txn_global->checkpoint_txn_shared.id) != WT_TXN_NONE) {
    if (id != txn->id)
       txn->snapshot[n++] = id;
    if (publish)
        txn_shared->metadata_pinned = id;
}
```

Reply:
Yep that makes sense, I wasn't too happy with the current solution.